server/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java

```diff
@@ -921,22 +923,50 @@ private <T> void termsAggregator(
                         .size(size)
                         .shardSize(size)
                         .field("field")
+                        .showTermDocCountError(true)
                         .order(bucketOrder);
 
-                    Terms result = searchAndReduce(
-                        indexSearcher,
-                        new AggTestConfig(aggregationBuilder, fieldType).withSplitLeavesIntoSeperateAggregators(false)
+                    AggTestConfig aggTestConfig = new AggTestConfig(aggregationBuilder, fieldType).withSplitLeavesIntoSeperateAggregators(
+                        // We cannot make result assertions on ascending order, so force exact counts by using only a single aggregator
+                        order == false
                     );
+                    Terms result = searchAndReduce(indexSearcher, aggTestConfig);
                     assertEquals(size, result.getBuckets().size());
+
+                    // Testing actual results, by first building a map of expected key->count because key order can change
+                    HashMap<Object, Long> expectedCounts = new HashMap<>();
+                    for (Map.Entry<T, Integer> expected : expectedBuckets) {
+                        Object key = (valueType == ValueType.IP) ? String.valueOf(expected.getKey()).substring(1) : expected.getKey();
+                        expectedCounts.put(key, expected.getValue().longValue());
+                    }
+
+                    // Compare actual counts to expected counts, taking into account potential errors due to per-shard doc limits
+                    long docCountError = result.getDocCountError();
                     for (int i = 0; i < size; i++) {
-                        Map.Entry<T, Integer> expected = expectedBuckets.get(i);
                         Terms.Bucket actual = result.getBuckets().get(i);
-                        if (valueType == ValueType.IP) {
-                            assertEquals(String.valueOf(expected.getKey()).substring(1), actual.getKey());
+                        long bucketDocCountError = actual.getDocCountError();
+                        assertThat(
+                            "Bucket doc count error is greater than the maximum doc count",
```

Comment:
This assertion tripped in ci. Maybe this can't always be asserted?

Reply:
When I run with `-Dtests.iters=1000` I get about 10 failures on this, so about 1% of the time. So this does seem to be an issue.